server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -4763,15 +4771,22 @@ public void syncSegmentsFromGivenRemoteSegmentStore(
         }
     }
 
+    // TODO: Move this method to a generic util class along-with the segment upload code
     private String copySegmentFiles(
         Directory storeDirectory,
         RemoteSegmentStoreDirectory sourceRemoteDirectory,
         RemoteSegmentStoreDirectory targetRemoteDirectory,
         Map<String, RemoteSegmentStoreDirectory.UploadedSegmentMetadata> uploadedSegments,
+        RemoteRefreshSegmentTracker downloadStatsTracker,
         boolean overrideLocal
     ) throws IOException {
+        boolean isRemoteStoreEnabled = indexSettings.isRemoteStoreEnabled();
```

Comment:
Why do we need to check if remote store is enabled here? Ideally, copySegmentFiles should be called only for remote store enabled indexes.

Reply:
Because we are using the same method in both recovery based downloads and snapshot interoperability. There is a possibility that while restoring snapshots (from a remote store) the end user explicitly disables remote store based settings (atleast that is the behavior up until now). In that case, for non-remote store indices the `RemoteStoreRefreshListener` is never executed, in turn not creating the stats tracker object.

Reply:
As I understand, `RemoteStoreRefreshListener` does not create `RemoteRefreshSegmentTracker`. Can you please check?

> There is a possibility that while restoring snapshots (from a remote store) the end user explicitly disables remote store based settings (atleast that is the behavior up until now).

I don't think this is possible. Please check.

Reply:
Sorry, my bad. It should be `RemoteRefreshSegmentPressureService` which is an `IndexEventListener`

https://github.com/opensearch-project/OpenSearch/blob/db838c4ca090103a8ba562b3cb124044be725c62/server/src/main/java/org/opensearch/index/remote/RemoteRefreshSegmentPressureService.java#L70-L85

This method is being used to create the tracker object. This was discovered when the following IT was run without the `remoteStoreEnabled` check. The IT trips if this check is not present:

https://github.com/opensearch-project/OpenSearch/blob/db838c4ca090103a8ba562b3cb124044be725c62/server/src/internalClusterTest/java/org/opensearch/snapshots/RestoreSnapshotIT.java#L289-L305